### PR TITLE
MAINT: stats: check integrality in `_argcheck` as needed

### LINKED ISSUE #2414: stats binom at non-integer n (Trac #1895)
_Original ticket http://projects.scipy.org/scipy/ticket/1895 on 2013-04-19 by @josef-pkt, assigned to @rgommers._

stats.binom does not impose that n is an integer.

The cdf seems to work correctly (?), the pmf doesn't sum to one.
The cdf from scipy.special might just floor the argument n (cast to int).
I don't know what to make of the case, when n is not an integer.
Is there an extension of the distribution for all real n>=0?

```
>>> stats.binom.cdf(np.arange(-3,5), 3.5, 0.5)
array([ 0.   ,  0.   ,  0.   ,  0.125,  0.5  ,  0.875,  1.   ,  1.   ])
>>> np.diff(stats.binom.cdf(np.arange(-3,5), 3.5, 0.5))
array([ 0.   ,  0.   ,  0.125,  0.375,  0.375,  0.125,  0.   ])
>>> stats.binom.pmf(np.arange(-3,5), 3.5, 0.5)
array([ 0.        ,  0.        ,  0.        ,  0.08838835,  0.30935922,
        0.38669902,  0.19334951,  0.        ])
>>> stats.binom.pmf(np.arange(-3,5), 3.5, 0.5).sum()
0.97779609585952287

>>> stats.binom.pmf(np.arange(-3,5), 3, 0.5)
array([ 0.   ,  0.   ,  0.   ,  0.125,  0.375,  0.375,  0.125,  0.   ])
```

### PR DESCRIPTION
**Pull request by mdhaber**
#### Reference issue
Closes gh-2414

#### What does this implement/fix?
Methods of some distributions return garbage when non-integral values are provided for shapes that are supposed to be integral. This PR makes existing discrete distribution methods return NaN in such cases. It also adds a test to protect new distributions from making the same mistake.

#### Additional information
I decided that parameters were "supposed to be integral" if:
* They are described in the documentation as though they should be integers
* SciPy has no tests for non-integral values
* There are no obvious references that suggest meaningful non-integer extensions (e.g. Wolfram Alpha)

According to gh-1110, `nbinom` is intended to do something meaningful with non-integral values of `n`. (NumPy's `negative_binomial` random variate generator, WolframAlpha, and Wikipedia all seem to agree that `n` can be any positive real.) When fitting this distribution with `scipy.stats.fit`, should we allow `n` to be any real?

Rather than adding explicit integrality checks to all `_argcheck` methods, at some point we should have a default `_argcheck` based on `_shape_info`. That will take some more work, so I decided to limit the scope. I think gh-2414 has been waiting long enough.

Non-integral values of `loc` can be confusing for these distributions, but not much more confusing than nonzero values of `loc`. I figure if there is a definition for nonzero values of `loc`, non-integral values of `loc` are fine, too.

### CODE REVIEW

File: scipy/stats/tests/test_discrete_basic.py [hunk 1]
```diff
@@ -319,3 +319,53 @@ def test_cdf_gh13280_regression(distname, args):
...
+    new_valid_shape = valid_shape - 1
+    shapes_copy[i] = [[valid_shape], [invalid_shape], [new_valid_shape]]
+
+    a, b = dist.support(*shapes)
+    x = np.round(np.linspace(a, b, 5))
+
+    pmf = dist.pmf(x, *shapes_copy)
+    assert not np.any(np.isnan(pmf[0, :]))
+    assert np.all(np.isnan(pmf[1, :]))
+    assert not np.any(np.isnan(pmf[2, :]))
```
Comment (mdhaber) [inline scipy/stats/tests/test_discrete_basic.py:371]: This makes sure the surrounding integral values are valid to ensure that it's the non-integrality that makes the values invalid.

This only checks the `pmf` method. Should we just make sure that `_argcheck` fails for non-integer values, or should we run this check on all methods?

**Comment by steppi** [comment]
It's been a very busy few weeks for me, but I should have time to look at this over the weekend. Sorry for the hold up.

**Comment by mdhaber** [comment]
NP, thanks @steppi!

**Review by steppi — APPROVED** [review]
I double-checked all of the distributions and agree with all of your choices for when to apply the integral check. I also think the test is fine as it is and doesn't need to check all methods. There's really only one minor change you may want to consider making. Lines [ [1074-1080]](https://github.com/scipy/scipy/blob/8141e152ee9a031674bb0df09930834163dddbff/scipy/stats/_distn_infrastructure.py#L1074-L1080) of `_distn_infrastructure.py`

```python
        args, loc, scale, size = self._parse_args_rvs(*args, **kwds)
        cond = logical_and(self._argcheck(*args), (scale >= 0))
        if not np.all(cond):
            message = ("Domain error in arguments. The `scale` parameter must "
                       "be positive for all distributions; see the "
                       "distribution documentation for other restrictions.")
            raise ValueError(message)  
```

now lead to a potentially confusing error message when `rvs` is called with non-integral arguments. At least I was a little confused by it at first when testing things out. Still approving though and deferring to you if you think it's worth updating the error message now.

Also, I'm +1 for allowing real `n` when fitting `nbinom`. `nbinom` is often used as an alternative to the Poisson distribution that allows the variance to be greater than the mean. Allowing real `n` is important for this application of fitting overdispersed count data.

**Comment by mdhaber** [comment]
Thanks @steppi I'll just update the error message to mention the other conditions on the input, I think.

As for fitting `nbinom` with real `n`, we'll have to think about how to handle that. I hesitate to add an argument to `scipy.stats.fit` that allows control over whether an arbitrary shape is constrained to be integral or not. (I imagine it would need to be controlled, right? Some users would still want `nbinom` fit only to integer `n`, right?)

It would be much easier if there were a separate distribution that was the same except for integrality of that shape parameter. We could document how users could subclass it, or we could actually make a separate distribution. What do you think of that? Any chance there's a different name for the distribution when `n` is not integral?

**Comment by mdhaber** [comment]
Updated the error message. Hopefully it's a bit better now?
```python3
from scipy import stats
stats.binom.rvs(10.5, 0.5)
```
```
ValueError: Domain error in arguments. The `scale` parameter must be positive for all distributions, 
and many distributions have restrictions on shape parameters. Please see the `scipy.stats.binom` 
documentation for details.
```
Even though `scale` is mentioned first, at least now it's more obvious that there could be a problem with a shape? 
If it looks good to you, please feel free to merge! And if you have a minute more to stop by gh-15723, I'd appreciate your throughts there.

**Comment by steppi** [comment]
> It would be much easier if there were a separate distribution that was the same except for integrality of that shape parameter. We could document how users could subclass it, or we could actually make a separate distribution. What do you think of that? Any chance there's a different name for the distribution when `n` is not integral? _Update: Polya?_

I actually wasn't aware of these alternative names

_The Pascal distribution (after [Blaise Pascal](https://en.wikipedia.org/wiki/Blaise_Pascal)) and Polya distribution (for [George Pólya](https://en.wikipedia.org/wiki/George_P%C3%B3lya)) are special cases of the negative binomial distribution. A convention among engineers, climatologists, and others is to use "negative binomial" or "Pascal" for the case of an integer-valued stopping-time parameter r, and use "Polya" for the real-valued case._

that are listed in the [Negative Binomial Distribution](https://en.wikipedia.org/wiki/Negative_binomial_distribution) Wikipedia article. I think the ideal nomenclature would be to have `nbinom` refer to the one with a real valued stopping parameter and `pascal` refer to one with the stopping parameter constrained to be an integer. I'd always seen "negative binomial distribution" refer to a distribution that allows for a real valued stopping parameter. For the sake of backwards compatibility though, I think it would be acceptable to use `polya` for the real-valued case and keep `nbinom`'s behavior the same.

I should have some time to look over at least one these other PRs tonight.

**Comment by mdhaber** [comment]
Well, the interest of backward compatibility is in your favor this time! 

`nbinom` has always accepted float `n`, so if we were to change things, it would be better to keep it that way and split off the one that accepts only integer `n` (`pascal`). `scipy.stats.fit` hasn't released yet, so there has been no way of fitting `nbinom` in the past - with integer `n` or otherwise. (And TBH, it would probably be considered a bug if we'd restricted `n` to integers when fitting, since real `n` has been OK in all the methods.)

I'll check with the mailing list!

**Comment by grlee77** [comment]
> nbinom has always accepted float n, so if we were to change things, it would be better to keep it that way and split off the one that accepts only integer n (pascal).

I saw the mailing list post about negative binomials and thought maybe this would be related to `scipy.special.nbdtr` for which I proposed deprecating non-integer `n` in https://github.com/scipy/scipy/pull/15845. However, looking at the code in `scipy.stats`, I see that `nbinom` as used in stats is from Boost and not `scipy.special`, so it is not directly related.

**Pull request merged**